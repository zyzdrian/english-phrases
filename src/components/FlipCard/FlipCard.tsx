import React, { useEffect, useState } from 'react';
import styled from 'styled-components';

export enum FLIP_CART_TEST_IDS {
    front = 'front',
    back = 'back'
}

const StyledFlip = styled.div`
perspective: 1000px;
z-index: auto;
`;

const StyledFlipper = styled.div`
position: relative;
width: 100%;
height: 100%;
`;

const FRONT_ANIMATION_START_ANGLE = 180;
const FRONT_ANIMATION_STOP_ANGLE = 0;

const StyledFront = styled.div<ISide>`
backface-visibility: hidden;
left: 0;
position: ${props => props.isFlipped ? 'absolute' : 'relative'};
top: 0;
transform: rotateY(${props => props.isFlipped ? FRONT_ANIMATION_START_ANGLE : FRONT_ANIMATION_STOP_ANGLE}deg);
transform-style: preserve-3d;
width: 100%;
height: 100%;
z-index: 2;
transition: all 0.6s ease 0s;
cursor: pointer;
`;

const BACK_ANIMATION_START_ANGLE = 0;
const BACK_ANIMATION_STOP_ANGLE = -180;

const StyledBack = styled.div<ISide>`
backface-visibility: hidden;
left: 0;
position: ${props => props.isFlipped ? 'relative' : 'absolute'};
transform: rotateY(${props => props.isFlipped ? BACK_ANIMATION_START_ANGLE : BACK_ANIMATION_STOP_ANGLE}deg);
transform-style: preserve-3d;
top: 0;
width: 100%;
height: 100%;
transition: all 0.6s ease 0s;
cursor: pointer;
`;

export const FlipCard: React.FC<IFlipCard> = ({ front, back, isOutsidedFlipped, onShow }) => {
    const [isFlipped, setFlipped] = useState(false);

    useEffect(() => {
        if (isOutsidedFlipped !== undefined) {
            setFlipped(isOutsidedFlipped);
        }
    }, [isOutsidedFlipped]);

    return (
        <StyledFlip>
            <StyledFlipper>
                <StyledFront
                    onClick={() => {
                        setFlipped(true);
                        onShow && onShow();
                    }}
                    isFlipped={isFlipped}
                    data-testid={FLIP_CART_TEST_IDS.front}
                >
                    {front}
                </StyledFront>
                <StyledBack
                    onClick={() => setFlipped(false)}
                    isFlipped={isFlipped}
                    data-testid={FLIP_CART_TEST_IDS.back}
                >
                    {back}
                </StyledBack>
            </StyledFlipper>
        </StyledFlip>
    );
};

export interface IFlipCard {
    front: React.ReactNode;
    back: React.ReactNode;
    isOutsidedFlipped?: boolean;
    onShow?: () => void;
}

interface ISide {
    isFlipped: boolean;
}
