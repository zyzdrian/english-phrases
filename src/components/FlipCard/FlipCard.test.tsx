import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { FlipCard, IFlipCard, FLIP_CART_TEST_IDS } from './FlipCard';

describe('<FlipCard /> tests', () => {
    const INITIAL_PROPS: IFlipCard = {
        back: <div>Back</div>,
        front: <div>Front</div>
    };
    let props = { ...INITIAL_PROPS };
    beforeEach(() => {
        props = { ...INITIAL_PROPS };
    });
    afterEach(cleanup);

    it('should render as default', () => {
        const { container, getByTestId } = render(<FlipCard {...props} />);
        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(FLIP_CART_TEST_IDS.front)).toHaveStyle('position: relative');
    });

    it('should render clickable front', () => {
        const { getByTestId } = render(<FlipCard {...props} />);
        fireEvent.click(getByTestId(FLIP_CART_TEST_IDS.front));

        expect(getByTestId(FLIP_CART_TEST_IDS.front)).toHaveStyle('position: absolute');
        // expect(getByTestId(FLIP_CART_TEST_IDS.front)).toHaveStyle('transform: rotateY(180deg)');
        expect(getByTestId(FLIP_CART_TEST_IDS.back)).toHaveStyle('position: relative');
        // expect(getByTestId(FLIP_CART_TEST_IDS.back)).toHaveStyle('transform: rotateY(0deg)');
    });

    it('should render clickable back', () => {
        const { getByTestId } = render(<FlipCard {...props} />);
        fireEvent.click(getByTestId(FLIP_CART_TEST_IDS.back));

        expect(getByTestId(FLIP_CART_TEST_IDS.front)).toHaveStyle('position: relative');
        // expect(getByTestId(FLIP_CART_TEST_IDS.front)).toHaveStyle('transform: rotateY(0deg)');
        expect(getByTestId(FLIP_CART_TEST_IDS.back)).toHaveStyle('position: absolute');
        // expect(getByTestId(FLIP_CART_TEST_IDS.back)).toHaveStyle('transform: rotateY(-180deg)');
    });

    it('should flip card when isOutsidedFlipped was changed', () => {
        props.isOutsidedFlipped = true;

        const { getByTestId } = render(<FlipCard {...props} />);

        expect(getByTestId(FLIP_CART_TEST_IDS.front)).toHaveStyle('position: absolute');
        expect(getByTestId(FLIP_CART_TEST_IDS.back)).toHaveStyle('position: relative');
    });

    it('should called onShow when it passed by props and click on front', () => {
        props.onShow = jest.fn();

        const { getByTestId } = render(<FlipCard {...props} />);
        fireEvent.click(getByTestId(FLIP_CART_TEST_IDS.front));

        expect(props.onShow).toBeCalled();
    });
});
