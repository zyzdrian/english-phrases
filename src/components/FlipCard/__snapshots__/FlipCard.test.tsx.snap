// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlipCard /> tests should render as default 1`] = `
<div
  class="sc-bdVaJa eZtYxm"
>
  <div
    class="sc-bwzfXH hhnQWX"
  >
    <div
      class="sc-htpNat gatBxD"
      data-testid="front"
    >
      <div>
        Front
      </div>
    </div>
    <div
      class="sc-bxivhb ggZgco"
      data-testid="back"
    >
      <div>
        Back
      </div>
    </div>
  </div>
</div>
`;
